### PR TITLE
Wx Toolbar for ToolManager

### PR DESCRIPTION
**Pull request by DietmarSchwertberger**
## PR Summary
This adds a first implementation of the ToolbarWx for use with ToolManager, including StatusbarWx, and ConfigureSubplotsWx.
The other tools were there already.
The ToolManager example is updated.

Additionally, an icon will be set also for FigureFrameWx.

I will also prepare an additional PR to replace the toolbar with a standard wx toolbar at the top of the frame. This will fix issue 2716 and make the FigureFrameWx initialization code more straightforward.

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_wx.py [hunk 1]
```diff
@@ -1697,9 +1741,107 @@ def set_function(self, string):
...
+        wx.ToolBar.__init__(self, parent, -1, style=style)
+        self._toolitems = {}
+        self._groups = {}
+        self._last = None
+
+    def add_toolitem(
+        self, name, group, position, image_file, description, toggle):
+
+        before, group = self._add_to_group(group, name, position)
+        idx = self.GetToolPos(before.Id)
```
Comment (fariza) [inline lib/matplotlib/backends/backend_wx.py]: where is this method defined? Pep8 lowercase

Comment (fariza) [reply lib/matplotlib/backends/backend_wx.py]: I guess it comes from wx.Toolbar, forget the previous comment

Comment (fariza) [reply lib/matplotlib/backends/backend_wx.py]: Sorry, it went out as review... I didn't know how to cancel the "start a review"

**Comment by fariza** [comment]
I have to install my wxpython correctly before I can continue the review. Soon...

**Comment by DietmarSchwertberger** [comment]
@fariza : That should not be too hard. If you're on Linux, then you should look here, as otherwise PIP will build from sources: https://wxpython.org/pages/downloads/

File: lib/matplotlib/backends/backend_wx.py [hunk 2]
```diff
@@ -1337,6 +1360,26 @@ def _load_bitmap(filename):
     return bmp
 
 
+_FRAME_ICON = None
```
Comment (fariza) [inline lib/matplotlib/backends/backend_wx.py]: I'm not that fond of this Gloabal icon.
It seems to be there just to save loading again the same image. 
Unless we are short on memory or it does impose a penalty on performance, I would say, just load it again

Comment (DietmarSchwertberger) [reply lib/matplotlib/backends/backend_wx.py]: It's actually a set of bitmaps with different sizes. As actually there are only two, I'm fine with reloading. I've just pushed the modification.

**Comment by fariza** [comment]
@anntzer can you help reviewing this?

**Comment by anntzer** [comment]
Would be much happier if (e.g.) https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_backends_interactive.py was modified to add testing for the tool manager (right now there's no CI at all on any of this).  Note that this could be done independently of (and before) this PR.

**Comment by DietmarSchwertberger** [comment]
For me, that's currently out of scope. I'm not familar enough with this yet.

**Comment by anntzer** [comment]
Needs a rebase.

**Comment by anntzer** [comment]
Let's just try to keep things moving and push back the discussion to when backend_tools becomes the default, if that ever happens...

**Comment by fariza** [comment]
There is a problem. I think it is in the doc building steps

```Warning, treated as error:
autodoc: failed to import module 'matplotlib.backends.backend_wxagg'; the following exception was raised:
Traceback (most recent call last):
  File "/home/circleci/.local/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 140, in import_module
    __import__(modname)
  File "/home/circleci/project/lib/matplotlib/backends/backend_wxagg.py", line 7, in <module>
    from .backend_wx import (
  File "/home/circleci/project/lib/matplotlib/backends/backend_wx.py", line 1795, in <module>
    class StatusbarWx(StatusbarBase, wx.StatusBar):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

**Comment by DietmarSchwertberger** [comment]
Unfortunately, I don't know how to interpret this message or how to fix it.
There are no metaclasses involved.
The inheritance structure for `StatusbarWx` is the same as e.g. for `ToolbarWx` which does not trigger the message:
```
class ToolbarWx(ToolContainerBase, wx.ToolBar):
    def __init__(self, toolmanager, parent, style=wx.TB_HORIZONTAL):
        ToolContainerBase.__init__(self, toolmanager)
        wx.ToolBar.__init__(self, parent, -1, style=style)
   ...

class StatusbarWx(StatusbarBase, wx.StatusBar):
    def __init__(self, parent, *args, **kwargs):
        StatusbarBase.__init__(self, *args, **kwargs)
        wx.StatusBar.__init__(self, parent, -1)
```

**Comment by anntzer** [comment]
That's due to the slightly atrocious mocking we use to make docs build even though wx is not installed (at https://github.com/matplotlib/matplotlib/blob/master/doc/sphinxext/mock_gui_toolkits.py#L111).  The solution is to remove the API docs for wx (similarly to https://github.com/matplotlib/matplotlib/pull/9993/files#diff-dd295241396a7490c85d37c53ae65dc2) and (not structly necessary, but may as well) remove the mocking mentioned above.
As argued elsewhere, the docs for the GUI integrations should be replaced by a single hand-written docs documenting the common API (and critical differences, when applicable) of all toolkits anyways...

**Comment by DietmarSchwertberger** [comment]
This change makes the test pass, but then the documentation would be gone, right?
Do you think that adding a `StatusBar` to `class MyWX(MagicMock)` would work?
That seems to be the difference between `ToolBar` and `StatusBar`.

**Comment by anntzer** [comment]
The doc build failure is https://github.com/matplotlib/matplotlib/pull/11107, please fetch, rebase, and push.
By the way, feel free to remove the "needs rebase" label after a rebase, makes it easier to keep track of PR status.

**Comment by DietmarSchwertberger** [comment]
I'm still not familiar with git and still looking for a better GUI tool. I probably should have used "fetch" instead of "pull". The commit history looks like a mess, but the code diff is correct.

This did re-trigger the ci/circleci, but they still fail....

**Comment by QuLogic** [comment]
You have extra commits because you did some weirdness when rebasing. There are actually _two_ copies of your commits now, one ending at e68cfac6ec70daa1220c322e9aa03785bc690013 that's the original, and one ending at a60100352f12916541a4a7521d7a27a1ecf50696 that was rebased yesterday.

When you rebase and try to push, git tells you to fetch & merge (or pull) first. While this is the safe option, it's also wrong in this case because it pulls the old copy of your commits. You _want_ to replace the old commits with the rebased ones, so you need to do `git push --force origin wxtoolmanager`.

This is not too difficult to fix (these instructions assume you have no other commits or changes):
```
git checkout wxmanager
git reset --hard a60100352f12916541a4a7521d7a27a1ecf50696
git push --force origin wxmanager
```

However, you have a pair of commits that do something and then immediately revert it; I would also take this opportunity to remove those. After the reset, do `git rebase --interactive upstream/master`, and your editor will popup with a list of commits. Remove the `avoid 'TypeError: metaclass` line and its revert, then exit the editor. Then git will replay your commits and leave out those two. You can follow up with the force-push after that.

**Comment by DietmarSchwertberger** [comment]
@QuLogic  thanks a lot. That looks much better now. I think the "--force" was the problem. I could not imagine that such a standard task would require such a flag. A nice GUI with preview etc. would be great...

**Comment by timhoffm** [comment]
Concerning `--force`: Well, you rewrote the history. This is potentially harmful if multiple people work on that branch. You should only ever rewrite the history if you're the only using the branch. `--force` is your explicitly consent that you know what you are doing.

**Comment by DietmarSchwertberger** [comment]
@QuLogic : as the changes are done, would you mind updating your review result? Or maybe even merge the PR? Once this and PR #11081 are merged, I would like to start work on Wayland support to fix issue https://github.com/matplotlib/matplotlib/issues/10417

**Comment by DietmarSchwertberger** [comment]
@tacaswell : no problem. Thanks for merging. Maybe the guidelines should be updated to encourage self merging when two reviewers have approved.
Anyway, I've been quite bit busy with other projects recently. After 3.0 I will create a PR for issue #2716 and then a PR for an improved wx backend including Windows Metafile support. This should be the basis for the discussion whether to keep or finally remove the wx backend.

**Pull request merged**